doc/js_tutorials/js_core/js_basic_ops/js_basic_ops.markdown

```diff
@@ -0,0 +1,380 @@
+    let B = src.data[row * src.cols * src.channels() + col * src.channels() + 2];
+    let A = src.data[row * src.cols * src.channels() + col * src.channels() + 3];
+}
+@endcode
+
+@note  Data manipulation is only valid for continuous Mat. You should use isContinuous() to check first.
+
+**2. at**
+
+Mat Type  | At Manipulation
+--------- | ---------------
+CV_8U     | ucharAt
+CV_8S     | charAt
+CV_16U    | ushortAt
+CV_16S    | shortAt
+CV_32S    | intAt
+CV_32F    | floatAt
+CV_64F    | doubleAt
+
+@code{.js}
+let row = 3, col = 4;
+let src = cv.imread("canvasInput");
+let R = src.ucharAt(row, col * src.channels());
+let G = src.ucharAt(row, col * src.channels() + 1);
+let B = src.ucharAt(row, col * src.channels() + 2);
+let A = src.ucharAt(row, col * src.channels() + 3);
+@endcode
+
+@note  At manipulation is only for single channel access and the value can't be modified.
+
+**3. ptr**
+
+Mat Type  | Ptr Manipulation | JavaScript Typed Array
+--------  | ---------------  | ----------------------
+CV_8U     | ucharPtr         | Uint8Array
+CV_8S     | charPtr          | Int8Array
+CV_16U    | ushortPtr        | Uint16Array
+CV_16S    | shortPtr         | Int16Array
+CV_32S    | intPtr           | Int32Array
+CV_32F    | floatPtr         | Float32Array
+CV_64F    | doublePtr        | Float64Array
+
+@code{.js}
+let row = 3, col = 4;
+let src = cv.imread("canvasInput");
+let pixel = src.ucharPtr(row, col);
+let R = pixel[0];
+let G = pixel[1];
+let B = pixel[2];
+let A = pixel[3];
+@endcode
+
+mat.ucharPtr(k) get the k th row of the mat. mat.ucharPtr(i, j) get the i th row and the j th column of the mat.
+
+Image ROI
+---------
+
+Sometimes, you will have to play with certain region of images. For eye detection in images, first
+face detection is done all over the image and when face is obtained, we select the face region alone
+and search for eyes inside it instead of searching whole image. It improves accuracy (because eyes
+are always on faces) and performance (because we search for a small area)
+
+We use the function: **roi (rect)**
+@param rect    rectangle Region of Interest.
+
+Try it
+------
+
+Try this demo using the code above. Canvas elements named roiCanvasInput and roiCanvasOutput have been prepared. Choose an image and
+click `Try it` to see the result. You can change the code in the textbox to investigate more.
+
+\htmlonly
+<!DOCTYPE html>
+<head>
+<style>
+canvas {
+    border: 1px solid black;
+}
+.err {
+    color: red;
+}
+</style>
+</head>
+<body>
+<div id="roiCodeArea">
+<h2>Input your code</h2>
+<button id="roiTryIt" disabled="true" onclick="roiExecuteCode()">Try it</button><br>
+<textarea rows="7" cols="80" id="roiTestCode" spellcheck="false">
+let src = cv.imread("roiCanvasInput");
+let dst = new cv.Mat();
+// You can try more different parameters
+let rect = new cv.Rect(100, 100, 200, 200);
+dst = src.roi(rect);
+cv.imshow("roiCanvasOutput", dst);
+src.delete(); dst.delete();
```

Comment:
The `delete` is emscripten convention for memory management. Please refer to https://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/embind.html#memory-management. Underline, if the JS object wrapping a smart pointer of OpenCV, it will invoke the `release` method when JS invokes the `delete`.

Does it make sense?

Reply:
Well, is there `release()` method? Can we call it instead with the same effect?
Looks like there is no native C++ code at all - all code is translated on JavaScript, right?

Reply:
There is no `release()` method exposed in JS objects. Right, all code is translated to JavaScript.